File: ui/src/teamTla.ts
/**
 * Must stay aligned with worldcup_pool/team_tla.py canonical_team_tla().
 * Used so saved advancer codes match checkbox / winner styling when feeds use alternate TLAs.
 */
const TEAM_TLA_CANONICAL: Record<string, string> = {
  CUR: "CUW",
  DEU: "GER",
  HOL: "NED",
  SCT: "SCO",
  URY: "URU", // football-data.org uses ISO alpha-3 URY; FIFA/pool uses URU
};

export function canonicalTeamTla(raw: string | null | undefined): string {
  const s = (raw || "").trim().toUpperCase();
  if (!s) return "";
  const clipped = s.length > 16 ? s.slice(0, 16) : s;
  return TEAM_TLA_CANONICAL[clipped] ?? clipped;
}

export function teamTlaEquals(a: string | null | undefined, b: string | null | undefined): boolean {
  const ca = canonicalTeamTla(a);
  const cb = canonicalTeamTla(b);
  return !!ca && ca === cb;
}

/** Map a saved champion code to a `<select>` option `value` when the roster uses an alternate TLA. */
export function winnerCodeForTeamSelect(
  saved: string | null | undefined,
  teams: readonly { code: string }[],
): string {
  const s = (saved ?? "").trim();
  if (!s) return "";
  if (!teams.length) return canonicalTeamTla(s) || s.toUpperCase();
  const up = s.toUpperCase();
  const exact = teams.find((t) => t.code.toUpperCase() === up);
  if (exact) return exact.code;
  const alias = teams.find((t) => teamTlaEquals(t.code, s));
  if (alias) return alias.code;
  const c = canonicalTeamTla(s);
  return c && c !== "?" ? c : s.toUpperCase();
}

File: worldcup_pool/team_tla.py
"""Normalize external team TLAs to FIFA-style codes used in WC2026 official draw data."""

from __future__ import annotations

# football-data.org / other feeds sometimes differ from Wikipedia/FIFA draw TLAs.
_TEAM_TLA_CANONICAL: dict[str, str] = {
    "CUR": "CUW",  # Curaçao — API often uses CUR; official draw / FIFA use CUW
    "DEU": "GER",  # ISO alpha-3 vs common football abbreviation
    "HOL": "NED",  # Historic Netherlands
    "SCT": "SCO",  # Scotland alternate in some feeds
    "URY": "URU",  # football-data.org uses ISO alpha-3 URY; FIFA/pool uses URU
}


def canonical_team_tla(raw: str | None, *, max_len: int = 16) -> str:
    """Uppercase TLA with known aliases folded to the canonical code."""
    s = (raw or "").strip().upper()
    if not s:
        return "?"
    if len(s) > max_len:
        s = s[:max_len]
    return _TEAM_TLA_CANONICAL.get(s, s)
